base/cvd/cuttlefish/host/commands/cvd/fetch/fetch_context.cc

```diff
@@ -0,0 +1,317 @@
+
+  std::string extracted_file = CF_EXPECT(
+      ExtractImage(archive_file, target_directory_, member_name, true));
+
+  std::string phase =
+      fmt::format("Extracted '{}' from '{}'", member_name, archive_name);
+  trace_.CompletePhase(std::move(phase), FileSize(extracted_file));
+
+  CF_EXPECT(fetch_context_.fetcher_config_.AddFilesToConfig(
+      file_source_, BuildId(), BuildTarget(), {extracted_file},
+      target_directory_, true));
+
+  CF_EXPECT(DesparseFiles({member_name}));
+
+  return {};
+}
+
+Result<void> FetchBuildContext::RenameFile(const std::string& from,
+                                           std::string_view to) {
+  std::string from_file = fmt::format("{}/{}", target_directory_, from);
+  std::string to_file = fmt::format("{}/{}", target_directory_, to);
+
+  if (std::string dir = android::base::Dirname(to_file); !dir.empty()) {
+    CF_EXPECT(EnsureDirectoryExists(dir, kRwxAllMode));
+  }
+
+  CF_EXPECT(::cuttlefish::RenameFile(from_file, to_file));
+
+  for (std::string dir = Dirname(from_file); IsDirectoryEmpty(dir);
+       dir = Dirname(dir)) {
+    CF_EXPECTF(rmdir(dir.c_str()) == 0, "rmdir failed: {}", StrError(errno));
+  }
+
+  CF_EXPECT(fetch_context_.fetcher_config_.AddFilesToConfig(
+      file_source_, BuildId(), BuildTarget(), {to_file}, target_directory_,
+      true));
+
+  CF_EXPECT(fetch_context_.fetcher_config_.RemoveFileFromConfig(from));
+
+  return {};
+}
+
+Result<void> FetchBuildContext::DeleteFile(const std::string& name) {
+  std::string filename = fmt::format("{}/{}", target_directory_, name);
+  CF_EXPECT(::cuttlefish::RemoveFile(filename));
+
+  CF_EXPECT(fetch_context_.fetcher_config_.RemoveFileFromConfig(name));
+
+  return {};
+}
+
+Result<ReadableZip> FetchBuildContext::OpenZip(const std::string& name) {
+  std::string local_path = fmt::format("{}/{}", target_directory_, name);
+  if (FileExists(local_path)) {
+    return CF_EXPECT(ZipOpenRead(local_path));
+  } else {
+    return CF_EXPECT(
+        ::cuttlefish::OpenZip(fetch_context_.build_api_, build_, name));
+  }
+}
+
+Result<void> FetchBuildContext::ExtractArchiveMember(
+    ReadableZip& zip, const std::string& member_name) {
+  std::string output = fmt::format("{}/{}", target_directory_, member_name);
+
+  if (std::string dir = Dirname(output); !dir.empty()) {
+    CF_EXPECT(EnsureDirectoryExists(dir, kRwxAllMode));
+  }
+
+  CF_EXPECTF(ExtractFile(zip, member_name, output), "Failed to extract '{}'",
+             member_name);
+
+  std::string phase = fmt::format("Extracted/Downloaded '{}'", member_name);
+  trace_.CompletePhase(std::move(phase), FileSize(output));
+
+  CF_EXPECT(fetch_context_.fetcher_config_.AddFilesToConfig(
+      file_source_, BuildId(), BuildTarget(), {output}, target_directory_,
+      true));
+
+  CF_EXPECT(DesparseFiles({member_name}));
+
+  return {};
+}
+
+std::string FetchBuildContext::BuildId() const {
+  return std::visit([](auto&& build) { return build.id; }, build_);
```

Comment:
nit: I see several calls to `std::visit()` with `cuttlefish::Build`, why is that preferred over making `cuttlefish::Build` the parent class of the other two?

Reply:
I won't claim `using Build = std::variant<DeviceBuild, DirectoryBuild>;` is a good API, but redesigning that would involve significant changes to [`android_build_api.cpp`](https://github.com/google/android-cuttlefish/blob/eaad16d8b41766ad48e9157ce603175a2551282e/base/cvd/cuttlefish/host/libs/web/android_build_api.cpp#L87) and this PR feels large enough as-is.

Reply:
I've replaced the `std::visit` calls with existing utility functions, i.e. `GetBuildIdAndTarget`, is this better?